### PR TITLE
Add new feature of SafeLoRA

### PR DESCRIPTION
**Pull request by chiayi-hsu**
The pull request was closed due to syncing with the latest version of PEFT, so I have requested the pull request again.
I have made all the necessary changes based on our previous conversations in this version.

If there are any issues, please let me know. 

Thank you.

### CODE REVIEW

File: examples/safelora/README.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# Safe LoRA 
+
+The official code of Safe LoRA:the Silver Lining of Reducing Safety Risks when Fine-tuning Large Language Models
+
```
Comment (BenjaminBossan) [inline examples/safelora/README.md:4]:
```suggestion
The official code of Safe LoRA: The Silver Lining of Reducing Safety Risks when Fine-tuning Large Language Models
```

File: examples/safelora/README.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+## Quick Start
+
+Please import the SafeLoraConfig and apply_safelora first.
```
Comment (BenjaminBossan) [inline examples/safelora/README.md:8]:
```suggestion
Please import the `SafeLoraConfig` and `apply_safelora` first.
```

File: examples/safelora/README.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+config = SafeLoraConfig(base_model_path='../LLM_Models/llama-2-7b-hf/',\
+                            aligned_model_path='../LLM_Models/llama-2-7b-chat-fp16/',
+                            peft_model_path = '../finetuneLLM/finetuned_models/samsumBad-7b-fp16-peft-seed-42',
+                            devices='cuda',
+                            select_layers_type='threshold',
+                            save_weights=True)
+
+final_lora_weight = apply_safelora(config)
+
```
Comment (BenjaminBossan) [inline examples/safelora/README.md:23]: Can we add a bit more to the example. For instance, how to save and load these weights?

File: examples/safelora/README.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+Please import the SafeLoraConfig and apply_safelora first.
+Then, fill in the paths for the base, aligned, and PEFT models according to your needs. There are two types of `select_layers_type`: `threshold` and `number`. The `threshold` type will determine how many layers will be projected based on the value you set. The `number` type directly specifies the number of projected layers.
+
+```python
+
+from peft.utils.safelora import SafeLoraConfig, apply_safelora
+
+config = SafeLoraConfig(base_model_path='../LLM_Models/llama-2-7b-hf/',\
+                            aligned_model_path='../LLM_Models/llama-2-7b-chat-fp16/',
```
Comment (BenjaminBossan) [inline examples/safelora/README.md:16]: Let's use the HF model ids for these two.

File: src/peft/utils/safelora.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        safelora_config.base_model_path, safelora_config.aligned_model_path, safelora_config.devices, peft_config, safelora_config
+    )
+
+    with safe_open(
+        f"{os.path.join(safelora_config.peft_model_path, 'adapter_model.safetensors')}", framework="pt", device=safelora_config.devices
+    ) as f:
+        if (safelora_config.devices).lower() == "cpu":
+            peft_weights = {name: f.get_tensor(name).to(safelora_config.dtype) for name in f.keys()}
+        else:
+            peft_weights = {name: f.get_tensor(name).to(safelora_config.dtype) for name in f.keys()}
```
Comment (BenjaminBossan) [inline src/peft/utils/safelora.py:217]: These 2 lines are identical

File: src/peft/utils/safelora.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    ori_peft_weights = copy.deepcopy(peft_weights)
+    vars_names_LoRA_A = [name for name in peft_weights.keys() if "lora_A" in name]
+    vars_names_LoRA_B = [name for name in peft_weights.keys() if "lora_B" in name]
+    num_projected_layers = 0
+    dis = []
+    cos_total = []
+    for idx, (name_A, name_B) in enumerate(zip(vars_names_LoRA_A, vars_names_LoRA_B)):
+        A = ori_peft_weights[name_A]
+        if configs.devices != "cpu":
+            P = v[idx].to(torch.bfloat16).to(configs.devices)
```
Comment (BenjaminBossan) [inline src/peft/utils/safelora.py:159]: Should we really hard-code bf16 here? Let's make it configurable. Also, I think with recent pytorch versions, bf16 works on CPU.

File: src/peft/utils/safelora.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+        else:
+            P = v[idx].to("cpu")
+        W = torch.mm(P, ori_peft_weights[name_B])
+        fW = torch.mm(W, A)
+        ori = torch.mm(ori_peft_weights[name_B], A)
+        cos = torch.round(torch.nn.functional.cosine_similarity(fW.reshape(1, -1), ori.reshape(1, -1)) * 10**5) / 10**5
```
Comment (BenjaminBossan) [inline src/peft/utils/safelora.py:165]: `torch.round` has a `decimals` argument, maybe use that instead of *10**5 (do we need the rounding at all?).

File: src/peft/utils/safelora.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+def get_aligned_matrix(base_model_path, aligned_model_path, devices, peft_config, configs):
+    """
+    Get projected matrix by following the config (target_modules) from the peft model.
+    The dimensions between the base model's weights and the aligned model's weights should be the same.
+    """
+    sl_align = SafetensorLoader(aligned_model_path, local_files_only=configs.local_files_only)
+    sl_base = SafetensorLoader(base_model_path, local_files_only=configs.local_files_only)
+
+    base_model_parameters = [
+        name for name in sl_base.weight_map.keys() if any(v in name for v in list(peft_config.target_modules))
```
Comment (BenjaminBossan) [inline src/peft/utils/safelora.py:129]: This check for `v in list(peft_config.target_modules)` is not always valid. E.g., `target_modules` can be a str in which case a regex match is executed. There can be other edge cases too. I think what should work is to use `LoraModel._check_target_module_exists(peft_config, name)` ([here](https://github.com/huggingface/peft/blob/b1fd97dc3e71d91abf9a8605e8be74ea6bb751c6/src/peft/tuners/lora/model.py#L159)).

File: src/peft/utils/safelora.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+    ]
+    align_model_parameters = [
+        name for name in sl_align.weight_map.keys() if any(v in name for v in list(peft_config.target_modules))
+    ]
```
Comment (BenjaminBossan) [inline src/peft/utils/safelora.py:133]: Should we also check that `base_model_parameters` and `align_model_parameters` are the same?

File: src/peft/utils/safelora.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+        vec = sl_base.get_tensor(name_base) - sl_align.get_tensor(name_align)
+        vec = vec.to(devices)
+        if devices == "cpu":
+            vec = vec.to(torch.float32)
+        vec = torch.mm(vec, vec.t()) / torch.norm(vec)
+        safety_vector.append((vec).detach().cpu())
+    return safety_vector
+
+
+def project_weights(configs, peft_weights, v):
```
Comment (BenjaminBossan) [inline src/peft/utils/safelora.py:149]: Let's rename `configs` to `config` or `safelora_config`.

File: src/peft/utils/safelora.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        default="TheBloke/Llama-2-7B-Chat-fp16",
+        metadata={"help": "The path of the aligned model for obtaining the aligned matrix."},
+    )
+
+    peft_model_path: str = field(
+        default="LisaSchunke/llama-2-7b-peft-finetuned-20000-dataset",
+        metadata={"help": "The path of the LoRA wieghts and configs."},
+    )
+
+    select_layers_type: str = field(
```
Comment (BenjaminBossan) [inline src/peft/utils/safelora.py:75]: Instead of `str`, we can annotate this as `Literal["threshold", "number"]`.

File: src/peft/utils/safelora.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+        default=0.5,
+        metadata={"help": "The threshold of cosine similarity for selecting projected layers."},
+    )
+
+    num_proj_layers: int = field(
+        default=10,
+        metadata={"help": "The number of projected layers."},
+    )
+
+    devices: str = field(
```
Comment (BenjaminBossan) [inline src/peft/utils/safelora.py:90]:
```suggestion
    device: str = field(
```

File: examples/safelora/safelora_inference.py [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+from peft.utils.safelora import SafeLoraConfig, apply_safelora
+
+config = SafeLoraConfig(base_model_path='../LLM_Models/llama-2-7b-hf/',\
+                            aligned_model_path='../LLM_Models/llama-2-7b-chat-fp16/',
+                            peft_model_path = '../finetuneLLM/finetuned_models/samsumBad-7b-fp16-peft-seed-42',
+                            devices='cuda',
+                            select_layers_type='threshold',
+                            save_weights=True)
+
+final_lora_weight = apply_safelora(config)
```
Comment (BenjaminBossan) [inline examples/safelora/safelora_inference.py:11]: The example should show inference, here we only create the weights. What are the next steps?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for the update to the SafeLoRA PR. I did another review and found a few areas to improve. Please take a look. Also, please run `make style` once you're finished with your changed.

File: src/peft/utils/safelora.py (hunk 6 above)
Comment (chiayi-hsu) [reply src/peft/utils/safelora.py:159]: Yes, you're right; bf16 does work on CPUs. I'll make it configurable.

File: examples/safelora/README.md (hunk 3 above)
Comment (chiayi-hsu) [reply examples/safelora/README.md:25]: I have added more descriptions to the example.
If you feel there are still any missing parts, please let me know.

File: examples/safelora/README.md (hunk 4 above)
Comment (chiayi-hsu) [reply examples/safelora/README.md]: Has been modified.

File: src/peft/utils/safelora.py (hunk 5 above)
Comment (chiayi-hsu) [reply src/peft/utils/safelora.py]: Has been modified.

```diff
- if (safelora_config.devices).lower() == "cpu":
-        peft_weights = {name: f.get_tensor(name).to(safelora_config.dtype) for name in f.keys()}
- else:
-        peft_weights = {name: f.get_tensor(name).to(safelora_config.dtype) for name in f.keys()}
+ peft_weights = {name: f.get_tensor(name).to(safelora_config.dtype) for name in f.keys()}
```

File: src/peft/utils/safelora.py (hunk 6 above)
Comment (chiayi-hsu) [reply src/peft/utils/safelora.py]: Has been modified.

```diff
- if configs.devices != "cpu":
-         P = v[idx].to(torch.bfloat16).to(configs.devices)
- else:
-         P = v[idx].to("cpu")
+ P = v[idx].to(safelora_config.dtype).to(safelora_config.device)

File: src/peft/utils/safelora.py (hunk 7 above)
Comment (chiayi-hsu) [reply src/peft/utils/safelora.py]: Has been modified.

```diff
- cos = torch.round(torch.nn.functional.cosine_similarity(fW.reshape(1, -1), ori.reshape(1, -1)) * 10**5) / 10**5
+ cos = torch.nn.functional.cosine_similarity(fW.reshape(1, -1), ori.reshape(1, -1))
```

File: src/peft/utils/safelora.py (hunk 9 above)
Comment (chiayi-hsu) [reply src/peft/utils/safelora.py:135]: I have added a check to verify if the model weights are the same.

```diff
+ if (sl_base.get_tensor(name_base) == sl_align.get_tensor(name_align)).all():
+        raise ValueError("The weights of the base Model and the aligned Model should be different.")
```

File: src/peft/utils/safelora.py (hunk 10 above)
Comment (chiayi-hsu) [reply src/peft/utils/safelora.py]: Has been modified.

File: src/peft/utils/safelora.py (hunk 11 above)
Comment (chiayi-hsu) [reply src/peft/utils/safelora.py]: Has been modified.

File: examples/safelora/safelora_inference.py (hunk 13 above)
Comment (chiayi-hsu) [reply examples/safelora/safelora_inference.py]: I have added more explanations in the README.md and also included code on how to use the SafeLoRA model.

File: src/peft/utils/safelora.py (hunk 8 above)
Comment (chiayi-hsu) [reply src/peft/utils/safelora.py:130]: I have tried to use ```LoraModel._check_target_module_exists(peft_config, name)```.
However, since the model here is not a PEFT model but a complete model, using its name to find the target module will result in not finding it

**Comment by BenjaminBossan** [comment]
@chiayi-hsu Once you're finished with your changes and want me to give another review, please ping me.

**Comment by chiayi-hsu** [comment]
@BenjaminBossan I have completed the modifications. Please help review them. Thanks!

File: src/peft/utils/safelora.py [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+from .loftq_utils import _SafetensorLoader as SafetensorLoader
+
+
+@dataclass
+class SafeLoraConfig:
+    """
+    This is the configuration class to store the configuration of a safeLora.
+
+
+    Args:
```
Comment (BenjaminBossan) [inline src/peft/utils/safelora.py:37]: Could you please format the docstring to be in line with the other docstrings used in PEFT. As an example, check here: https://github.com/huggingface/peft/blob/8874ab5ed8886f988855c93cedd0f323724d35f7/src/peft/tuners/lora/config.py#L128

File: src/peft/utils/safelora.py [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+        metadata={"help": "The path of the base model for obtaining the aligned matrix."},
+    )
+
+    aligned_model_path: str = field(
+        default="TheBloke/Llama-2-7B-Chat-fp16",
+        metadata={"help": "The path of the aligned model for obtaining the aligned matrix."},
+    )
+
+    peft_model_path: str = field(
+        default="LisaSchunke/llama-2-7b-peft-finetuned-20000-dataset",
```
Comment (BenjaminBossan) [inline src/peft/utils/safelora.py:72]: IMO, it doesn't make sense to set default values here, I would remove them. WDYT?

File: src/peft/utils/safelora.py (hunk 9 above)
Comment (BenjaminBossan) [reply src/peft/utils/safelora.py:135]: I meant something else. Would we expect that `base_model_parameters == align_model_parameters`? If not, under what circumstances would they differ?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks a lot for the updates. I did another review. Most of what I found are just smaller things like docs, please take a look.

Now as a next step, it is important that we also add some unit tests. This not going to be very straightforward, because we cannot easily test model alignment and we also don't want to use any big models during unit testing.

One proposal for this would be to use a small model like `hf-internal-testing/tiny-random-OPTForCausalLM` as the base model. Then let's modify some weights (setting them to 0?) and save this as the "aligned" model. Then call `apply_safelora` with these 2 models and various options to see if those tests pass. This would not really check the alignment though.

In addition, we could think about adding a true alignment test for the nightly run with GPU. For this test, it would be okay to use a bigger model (but ideally still not too big).

LMK what you think about this testing strategy and if you have further questions.

Apart from this, please call `make style` on your PR, as this is a prerequisite for the CI to pass.

File: src/peft/utils/safelora.py [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+        name for name in sl_align.weight_map.keys() if any(v in name for v in list(peft_config.target_modules))
+    ]
+    safety_vector = []
+    for name_base, name_align in zip(base_model_parameters, align_model_parameters):
+        if sl_base.get_tensor(name_base).shape != sl_align.get_tensor(name_align).shape:
+            raise ValueError(
+                "The dimensions of the base model's weight should be the same with the aligned model's weight."
+            )
+        if (sl_base.get_tensor(name_base) == sl_align.get_tensor(name_align)).all():
+            raise ValueError("The weights of the base Model and the aligned Model should be different.")
```
[the remark this answers is not in the document]
Comment (chiayi-hsu) [reply src/peft/utils/safelora.py:143]: In our setting, the base model refers to a model that has not undergone training with techniques such as RLHF or SFT, while the aligned model refers to one that has been trained using these techniques. Currently, open-source models such as LLaMA 2 or 3, Mistral, and Gemma have released both base models and chat/instruct (aligned) models. Therefore, we expect the weights of the base model and the aligned model to differ.

**Comment by chiayi-hsu** [comment]
Yes, it still needs to be addressed.

github-actions[bot] ***@***.***>於 2024年12月27日 週五，16:04寫道：

**Comment by chiayi-hsu** [comment]
@BenjaminBossan 
I have made the modifications based on your review and ensured that the style passes by running make style. Regarding the unit test, we can follow your suggestion to use a small model like hf-internal-testing/tiny-random-OPTForCausalLM and modify part of the weights in each layer to be 0, treating it as the so-called aligned model. In our setting, we require that the weights of the base and aligned models be entirely different because we want the aligned model to be a fully fine-tuned model obtained through RLHF techniques. This ensures that the aligned matrix obtained is more flexibly applicable to various attention layers where LoRA is applied.

I would like to ask about the unit test part. Do I need to write test scripts for this? Are there any specific rules or things I should pay attention to?

File: examples/safelora/README.md [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+
+```
+### Save SafeLoRA's Weights
+If you set `save_weights=False`, but still want to save the weights, you can use the following code.
+
+```python
+from safetensors.torch import save_file
+
+path = ...  # your PEFT model path
+save_file(final_lora_weight, os.path.join(path, "adapter_model.safetensors"))
```
Comment (BenjaminBossan) [inline examples/safelora/README.md:34]: Here, `path` would be the same as `peft_path` above, right? If so, let's use the same name.

File: examples/safelora/safelora_inference.py [hunk 18]
```diff
@@ -0,0 +1,28 @@
+import os
```
Comment (BenjaminBossan) [inline examples/safelora/safelora_inference.py:1]: Let's add the copyright notice.

File: examples/safelora/README.md [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+### Use SafeLoRA Model
+Next, you can load the base model of the Peft Model along with the Peft model itself to use a model that has both downstream task utility and alignment.
+
+```python
+from transformers import AutoModelForCausalLM
+from peft import PeftConfig, PeftModel
+
+model = AutoModelForCausalLM.from_pretrained(<base-model-id>)
+model = PeftModel.from_pretrained(model, <SafeLoRA-path>)
+```
```
Comment (BenjaminBossan) [inline examples/safelora/README.md:46]: There is no mention of the `safelora_inference.py` script in this README. Let's mention this script with a short explanation of what it does.

File: examples/safelora/safelora_inference.py [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+    select_layers_type="threshold",
+    save_weights=True,
+)
+
+final_lora_weight = apply_safelora(config)
+
+save_file(
+    final_lora_weight,
+    f"{os.path.join('../finetuneLLM/finetuned_models/samsumBad-7b-fp16-peft-seed-42', 'adapter_model.safetensors')}",
+)
```
Comment (BenjaminBossan) [inline examples/safelora/safelora_inference.py:25]: Can this be removed since we have already set `save_weights=True` above?

File: src/peft/utils/safelora.py [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+
+from .loftq_utils import _SafetensorLoader as SafetensorLoader
+
+
+@dataclass
+class SafeLoraConfig:
+    """
+    This is the configuration class to store the configuration of a [`safeLora`].
+
+
```
Comment (BenjaminBossan) [inline src/peft/utils/safelora.py:37]:
```suggestion
```

File: src/peft/utils/safelora.py [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+import peft
+from peft import PeftConfig
+
+from .loftq_utils import _SafetensorLoader as SafetensorLoader
+
+
+@dataclass
+class SafeLoraConfig:
+    """
+    This is the configuration class to store the configuration of a [`safeLora`].
```
Comment (BenjaminBossan) [inline src/peft/utils/safelora.py:35]: Regarding [`safeLora`]: There is no object of this name, right? Therefore, this syntax will not work. We could rephrase this to:

> This is the configuration class to set the parameters for SafeLoRA.

Then add a link to the paper.

File: src/peft/utils/safelora.py [hunk 23]
```diff
@@ -0,0 +1,227 @@
...
+    )
+
+    aligned_model_path: str = field(
+        default="TheBloke/Llama-2-7B-Chat-fp16",
+        metadata={"help": "The path of the aligned model for obtaining the aligned matrix."},
+    )
+
+    peft_model_path: str = field(
+        default="LisaSchunke/llama-2-7b-peft-finetuned-20000-dataset",
+        metadata={"help": "The path of the LoRA wieghts and configs."},
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply src/peft/utils/safelora.py:86]: Typo is still there.

File: src/peft/utils/safelora.py [hunk 24]
```diff
@@ -0,0 +1,257 @@
...
+        },
+    )
+
+    def __post_init__(self):
+        if self.base_model_path is None:
+            raise ValueError("base_model_path cannot be None.")
+        if self.aligned_model_path is None:
+            raise ValueError("aligned_model_path cannot be None.")
+        if self.peft_model_path is None:
+            raise ValueError("peft_model_path cannot be None.")
```
Comment (BenjaminBossan) [inline src/peft/utils/safelora.py:131]: Since there are no longer any default values for these fields, I don't believe we need to perform these checks anymore.

File: src/peft/utils/safelora.py (hunk 9 above)
Comment (BenjaminBossan) [reply src/peft/utils/safelora.py:148]: Still open.

File: src/peft/utils/safelora.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply src/peft/utils/safelora.py:156]: I understand the difference between aligned model and base model. However, it should be possible to align a model without changing each and every parameter, right? E.g., in the future we could a new model where the aligned model only changes half of the weights, I don't see why that couldn't be possible. Does the SafeLoRA algorithm really require each parameter to be different? Can we not skip the layers if the parameters are identical?

File: src/peft/utils/safelora.py [hunk 25]
```diff
@@ -0,0 +1,257 @@
...
+
+        dist = 1 / (1 + torch.norm(peft_weights[name_B].reshape(1, -1) - W.reshape(1, -1)))
+
+        dis.append(dist.item())
+    return peft_weights, cos_total
+
+
+def apply_safelora(safelora_config: SafeLoraConfig):
+    """
+
```
Comment (BenjaminBossan) [inline src/peft/utils/safelora.py:193]:
```suggestion
```

File: src/peft/utils/safelora.py [hunk 26]
```diff
@@ -0,0 +1,257 @@
...
+    Therefore, applying `apply_safelora()` is intended to help preserve the alignment of the final models.
+
+    It is important to note that the model weights of the aligned model and the base model must be of the same size.
+    Additionally, for SafeLoRA, only the safetensors format is supported.
+
+    Args:
+        safelora_config: The config of SafeLora.
+
+    Returns:
+        `torch.nn.Module`: The Lora model is applied SafeLoRA.
```
Comment (BenjaminBossan) [inline src/peft/utils/safelora.py:207]: This is incorrect, the return value is a `state_dict` containing the PEFT weights.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thank you for the updates.

I reviewed the PR again and found a few more things that can be improved, it's mostly about clarity and formatting.

Regarding the unit tests:

I think it would be easiest to proceed as follows. Let's create a new file, `tests/test_safelora.py`. Next, I created a small template for you to get started:

```python
class TestSafeLora:
    model_id = "hf-internal-testing/tiny-random-OPTForCausalLM"

    @pytest.fixture(scope="class")
    def aligned_model_path(self, tmp_path):
        # we create a fake aligned model where the weights differ from the base model
        model = AutoModelForCausalLM.from_pretrained(self.model_id)
        for param in model.parameters():
            # modify the parameters to be different
            ...

        model.save_pretrained(tmp_path / "aligned_model")

    @pytest.fixture
    def lora_path(self, tmp_path):
        # create a LoRA adapter
        model = AutoModelForCausalLM.from_pretrained(self.model_id)
        lora_config = LoraConfig(init_lora_weights=False)  # initialize LoRA so that it's not a no-op
        model = get_peft_model(model, lora_config)
        model.save_pretrained(tmp_path / "lora")

    def test_safelora_with_threshold(self, aligned_model_path, lora_path):
        ...  # code to test

    def test_safelora_with_num_proj_layers(self, aligned_model_path, lora_path):
        ...  # code to test

    def test_safelora_with_save_weights_false(self, aligned_model_path, lora_path):
        ...  # code to test

    # etc. more tests for the different options here
```

I hope this makes sense, if not, feel free to ask questions.

As discussed earlier, since this does not use a real aligned model but just creates a fake one, we cannot really test if the final model is better aligned or not. For this, we rely on the paper results and assume they're correct. If you have a good idea for a real alignment test, we can also add that test though. Just ensure that we're only using very small models to not slow down the CI.

**Comment by chiayi-hsu** [comment]
I'm still working on it, but due to an upcoming paper submission
deadline, there will be some delays.

Benjamin Bossan ***@***.***> 於 2025年2月10日 週一 上午11:54寫道：

**Comment by BenjaminBossan** [comment]
Hey @chiayi-hsu I was recently thinking about this PR and wondered if you're interested in reviving it.
